### PR TITLE
[Merged by Bors] - test: Rewrite test-crate-version in Rust

### PR DESCRIPTION
**Pull request by mfdorst**
**Notable changes**
+ Does not download `https://static.crates.io/db-dump.tar.gz` to check if crates are published (very slow)
   + Instead requests `https://crates.io/api/v1/crates/{crate_name}/versions` - 404 indicates that the crate is not published

**TODO**
+ [x] Test source diffing
+ [x] Display source changes
+ [x] Test manifest diffing
  + [x] Decide whether deleted manifest keys should be ignored or flagged
+ [x] Display manifest changes
+ [x] Download crates without an external tool (currently using `cargo-download`)
+ [ ] ~~Diff source without an external tool (currently using `diff`) - maybe? Might not be worth the effort.~~
+ [x] Add a `--verbose` mode
+ [x] Download crates concurrently
+ [x] Colorize diff

Related: #2590

### CODE REVIEW

**Comment by mfdorst** [comment]
Thinking of using [treediff] in conjunction with [cargo-manifest] for showing changes in manifest files.
[treediff]: https://docs.rs/treediff/4.0.2/treediff/
[cargo-manifest]: https://docs.rs/cargo-manifest/0.3.0/cargo_manifest/

**Comment by mfdorst** [comment]
So an interesting thing about the current state of the manifest diffing:
+ Reordering keys shows as no change (good!)
+ Deleting a key also shows not change (bad! or maybe potentially okay?)
I honestly don't know if this is a problem. Removing an unused dependency should not require a version bump (I think?) but there are probably other keys that ought to require a version bump if removed. But maybe this is a rare enough thing that we don't care? Feedback would be appreciated.

**Comment by mfdorst** [comment]
I couldn't find a good crate for diffing toml files, so I'm making my own. [`toml-diff`]

[`toml-diff`]: https://github.com/mfdorst/toml-diff

**Comment by sehz** [comment]
https://github.com/infinyon/k8-api/tree/master/src/k8-diff

**Comment by tjtelan** [comment]
> So an interesting thing about the current state of the manifest diffing:
> 
>     * Reordering keys shows as no change (good!)
> 
>     * Deleting a key also shows no change (bad! or maybe potentially okay?)
> 
> 
> I honestly don't know if this is a problem. It may even be desirable. Removing an unused dependency should not require a version bump (I think?) but there are probably other keys that ought to require a version bump if removed. But maybe this is a rare enough thing that we don't care? Feedback would be appreciated.

The desire to have a version bump is primarily driven by our release process. Since we [use path dependencies](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-path-dependencies) for the public crates in the repo during development and testing, we have expectations that the public version is compatible after release.

So if we could be objective in order to decide what changes require or don't require a new version bump, and cover our internal dependency tree, then that sounds ok to me. As long as we're consistent between our internal dependencies.

**Comment by mfdorst** [comment]
> https://github.com/infinyon/k8-api/tree/master/src/k8-diff

There might be a way to get that working for toml, but I wasn't able to easily figure it out. Good news is my `toml-diff` crate is coming along nicely - it's working for basic toml, though it could use some improvement on nested arrays/tables.

**Comment by mfdorst** [comment]
I wasn't sure if I should include the crate in the source, or if I should publish it myself and have it pull from crates.io.

**Comment by mfdorst** [comment]
Yes it is.

**Comment by sehz** [comment]
Great.  Thanks for the nice work!
@tjtelan Please provide feedback

File: release-tools/check-crate-version/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+use serde::Deserialize;
+use toml_diff::TomlDiff;
+
+mod download;
+
+use download::download_crate;
+
+const PUBLISH_LIST_PATH: &str = "./publish-list.toml";
+const CRATES_DIR: &str = "../../crates";
+const CRATES_IO_DIR: &str = "./crates_io";
```
Comment (tjtelan) [inline release-tools/check-crate-version/src/main.rs:13]: These paths can keep those default values, but they should be overridable via CLI flags and environment variables. I believe Clap will handle both.

File: release-tools/check-crate-version/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+    let list_toml = fs::read_to_string(PUBLISH_LIST_PATH).unwrap();
+    let list: PublishList = toml::from_str(&list_toml).unwrap();
+    list.publish_list
+}
+
+/// Checks crates.io to see if the crate has been published. Returns `true` if it has.
+async fn check_crate_published(crate_name: &str) -> bool {
+    // We have to identify ourselves, or crates.io will ignore our request
+    // TODO: Add an email to the user_agent filed - this will make us less likely to have our
```
Comment (tjtelan) [inline release-tools/check-crate-version/src/main.rs:117]: you can use `team@infinyon.com` for crates.io's crawler policy

File: release-tools/check-crate-version/toml-diff/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+[package]
+name = "toml-diff"
+version = "0.0.0"
```
Comment (tjtelan) [inline release-tools/check-crate-version/toml-diff/Cargo.toml:3]: Can you also add `publish = false`? I don't think we intend to publish this crate at this time. At least not while it is living in this repo

File: release-tools/check-crate-version/toml-diff/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+#[derive(Debug)]
+pub enum TomlChange<'a> {
+    Added(Vec<&'a str>, &'a TomlValue),
+    Deleted(Vec<&'a str>, &'a TomlValue),
+}
+
+impl<'a> TomlDiff<'a> {
+    /// Return a list of differences between `a` and `b`. A is considered "new" and `b` is
+    /// considered "old", so items missing from `a` are considdered "deletions", while items
```
Comment (tjtelan) [inline release-tools/check-crate-version/toml-diff/src/lib.rs:22]: Small nitpick, a few typos of the same word this this doc comment.

`considdered` -> `considered`

File: release-tools/check-crate-version/toml-diff/src/test.rs [hunk 5]
```diff
@@ -0,0 +1,245 @@
+use super::{TomlChange, TomlDiff};
```
Comment (tjtelan) [inline release-tools/check-crate-version/toml-diff/src/test.rs:1]: These tests look good, but can you add a control case where there is no diff

File: release-tools/check-crate-version/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,18 @@
+[package]
+name = "check-crate-version"
+version = "0.1.0"
```
Comment (tjtelan) [inline release-tools/check-crate-version/Cargo.toml:3]: Can you update the version of this crate to `0.0.0` and add `publish = false`? If decide to publish this, it most likely won't be from this repo.

File: release-tools/check-crate-version/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    let padding = publish_list
+        .iter()
+        .fold(0, |len, name| max(len, name.len()));
+
+    let status_checks: Vec<_> = publish_list
+        .into_iter()
+        .map(|crate_name| {
+            (
+                crate_name.clone(),
+                tokio::spawn(check_crate_status(crate_name)),
```
Comment (tjtelan) [inline release-tools/check-crate-version/src/main.rs:44]: Can we add a rate limit for these checks so we can abide by crates.io's crawler policy of `1 request per second or less`?

https://crates.io/policies#crawlers

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -297,14 +297,14 @@ jobs:
         run: make run-all-unit-test
       - name: Crate version check
         if: matrix.check == 'crate-version'
-        run: ./release-scripts/test-crate-version.sh
+        run: ./release-tools/test-crate-version.sh
```
Comment (tjtelan) [inline .github/workflows/ci.yml:300]: You should run your tool to perform this check

**Review by tjtelan — COMMENTED** [review]
This looks like it took a lot of effort and will be a lot easier to maintain. Good job with `toml-diff`!

File: release-tools/check-crate-version/toml-diff/src/lib.rs (hunk 4 above)
Comment (mfdorst) [reply release-tools/check-crate-version/toml-diff/src/lib.rs:22]: I don't think I copied those lines, I think my fingers think it's spelled like that lol

File: release-tools/check-crate-version/Cargo.toml (hunk 6 above)
[the remark this answers is not in the document]
Comment (mfdorst) [reply release-tools/check-crate-version/Cargo.toml]: Fixed

File: release-tools/check-crate-version/src/main.rs (hunk 1 above)
Comment (mfdorst) [reply release-tools/check-crate-version/src/main.rs]: Implemented

File: release-tools/check-crate-version/src/main.rs (hunk 2 above)
Comment (mfdorst) [reply release-tools/check-crate-version/src/main.rs]: Implemented

File: release-tools/check-crate-version/toml-diff/Cargo.toml (hunk 3 above)
Comment (mfdorst) [reply release-tools/check-crate-version/toml-diff/Cargo.toml:3]: Fixed

File: release-tools/check-crate-version/toml-diff/src/lib.rs (hunk 4 above)
Comment (mfdorst) [reply release-tools/check-crate-version/toml-diff/src/lib.rs]: Fixed

File: release-tools/check-crate-version/toml-diff/src/test.rs (hunk 5 above)
Comment (mfdorst) [reply release-tools/check-crate-version/toml-diff/src/test.rs:1]: Added

File: release-tools/check-crate-version/Cargo.toml (hunk 6 above)
Comment (mfdorst) [reply release-tools/check-crate-version/Cargo.toml]: Fixed

File: release-tools/check-crate-version/src/main.rs (hunk 7 above)
Comment (mfdorst) [reply release-tools/check-crate-version/src/main.rs]: Bummer, cause it means there was no point in making this async, but you're right, that is what the crawler policy says. Implemented.

File: .github/workflows/ci.yml (hunk 8 above)
Comment (mfdorst) [reply .github/workflows/ci.yml]: Implemented

**Review by tjtelan — APPROVED** [review]
LGTM, I really appreciate the hard work that went into this

File: release-tools/check-crate-version/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,19 @@
+[package]
+name = "check-crate-version"
+version = "0.0.0"
+publish = false
+edition = "2021"
+
```
Comment (sehz) [inline release-tools/check-crate-version/Cargo.toml:6]: can you add a license?  this should be apache2

File: release-tools/check-crate-version/toml-diff/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+publish = false
+edition = "2021"
+
```
Comment (sehz) [inline release-tools/check-crate-version/toml-diff/Cargo.toml:6]: please add license

**Review by sehz — CHANGES_REQUESTED** [review]
Overall looks reasonable.  Just a couple of minor nits regarding the license.

File: release-tools/check-crate-version/Cargo.toml (hunk 9 above)
Comment (mfdorst) [reply release-tools/check-crate-version/Cargo.toml:7]: Done

File: release-tools/check-crate-version/toml-diff/Cargo.toml (hunk 10 above)
Comment (mfdorst) [reply release-tools/check-crate-version/toml-diff/Cargo.toml:7]: Done

**Review by sehz — APPROVED** [review]
LGTM.   This is a fairly good effort.  Thanks for working on this.

**Pull request closed without merging**